test/functional/sapling_wallet.py

```diff
@@ -166,17 +166,39 @@ def run_test(self):
         self.nodes[2].generate(1)
         self.sync_all()
 
+        # Shield more funds to a different address to verify multi-source notes spending
+        saplingAddr2 = self.nodes[0].getnewshieldedaddress()
+        self.log.info("TX 7: shield funds to later verify multi source notes spending.")
+        recipients = [{"address": saplingAddr2, "amount": Decimal('10')}]
+        mytxid7 = self.nodes[0].shieldedsendmany(get_coinstake_address(self.nodes[0]), recipients, 1, fee)
+        self.check_tx_priority([mytxid7])
+
+        self.nodes[2].generate(5)
+        self.sync_all()
+
+        # Verify multi-source notes spending
+        tAddr0 = self.nodes[0].getnewaddress()
+        self.log.info("TX 8: verifying multi source notes spending.")
+        recipients = [{"address": tAddr0, "amount": Decimal('11')}]
+        mytxid8 = self.nodes[0].shieldedsendmany("from_shielded", recipients, 1, fee)
+        self.check_tx_priority([mytxid8])
+
+        self.nodes[2].generate(1)
+        self.sync_all()
+
         # Verify balance
-        assert_equal(self.nodes[0].getshieldedbalance(saplingAddr0), Decimal('7'))   # 30 received - (20 sent + 3 fee)
+        assert_equal(self.nodes[0].getshieldedbalance(saplingAddr0), Decimal('3'))   # 30 received - (20 sent + 3 fee) - 7 sent + 3 change
         assert_equal(self.nodes[1].getshieldedbalance(saplingAddr1), Decimal('20'))  # 20 received
+        assert_equal(self.nodes[0].getshieldedbalance(saplingAddr2), Decimal('2'))   # 10 received - 10 sent + 2 change
```

Comment:
The change should not be split to two addresses.
Either the tx is creating an extra shielded output (and it shouldn't), or the shielded balance is not being updated properly.

We can fix this in another PR, but I would keep the test failing, rather than adjust it to make it pass (hiding the issue).
Or at least putting a huge comment there.

Reply:
Yes, i know. I'm investigating the problem.
I made the adaptations here because the issue isn't a problem introduced in this PR, it's an issue in our codebase. I tried to focus this PR on fixing, and there by enabling, multi-source notes spending which were crashing before.

But.. I agree that can work on it here as well and don't adapt the test.

Reply:
ok no, the code is fine. Just understood what is going on.

Essentially at the moment of spending tx8, we have the following context:

`node0` has two shielded addresses `saplingAddr0` and `saplingAddr2`. With the following notes distribution:
`saplingAddr0` has two available notes: note_1 with 4 PIV, note_2 with 3 PIV.
`saplingAddr2` has one available note: note_3 with 10 PIV.

So.. as tx8 is spending 11 shielded PIV with 1 PIV fee, the automatic notes selection takes note_1 and note_3 as inputs (because notes are sorted in descending order by amount), having a total of 14 shielded PIVs.

Now the clarity moment:
11 PIVs are going to the transparent address, 1 PIV to the fee and the 2 remaining PIVs are going to the change addr, creating note_4, that is being created from the **first** note of the transaction (the bigger one in terms of amount).

So.. it's fine that `saplingAddr0` finishes the process with 3 PIVs as it only has note_2 available and it's fine that `saploingAddr2` finishes the process with 2 PIV as it received note_4 on tx8.

The issue was that we were, badly, assuming that the 7 PIVs that `saplingAddr0` had were all on a single note. The created change is fine.

Reply:
Ok, yeah, makes sense.
We just need to update the first comment there.
```
 # 30 received - (20 sent + 3 fee) - 4 sent
```

Reply:
yep, doing it with the rebase.